### PR TITLE
feat(queuev2): add InMemQueue

### PR DESCRIPTION
**Pull request by arzonus**
**What changed?**

Adds `InMemQueue`, a sorted in-memory task store backed by a `[]persistence.Task` slice and binary-searched on `persistence.HistoryTaskKey`. Provides `PutTasks`, `LookAHead`, `GetTasks`, `LTrim`, `RTrimBySize`, `Clear`, and `Len`. Not internally synchronized — callers own the lock.

**Why?**

The upcoming `CachedQueueReader` needs a bounded, sorted buffer to hold lookahead tasks in memory between persistence reads. `InMemQueue` is that buffer: it keeps tasks ordered by key, supports efficient range reads with predicate filtering and pagination, and can be trimmed from either end to enforce size limits.

Part of changes implementing Timer Task Read Optimization https://github.com/cadence-workflow/cadence/issues/7953.

**How did you test it?**

Unit tests with the race detector:

```
go test -race ./service/history/queuev2/...
```

628-line table-driven test suite covering: empty-queue edge cases, single/multi-task inserts, out-of-order and duplicate-key handling for `PutTasks`; `LookAHead` at-before-after-boundary cases; `GetTasks` pagination, predicate filtering, boundary exclusion, zero/negative page sizes; `LTrim` left-boundary semantics; `RTrimBySize` with zero, negative, exact, and over-size inputs; and `Clear`/`Len` invariants.

**Potential risks**

N/A — new internal type, no public API or schema changes. Not wired into any production path yet.

**Release notes**

N/A — internal building block for the cached queue reader, not user-facing.

**Documentation Changes**

N/A

### CODE REVIEW

File: service/history/queuev2/queue_mem.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+				// Collected one extra to detect that a next page exists;
+				// stop scanning to avoid unnecessary allocations.
+				break
+			}
+		}
+	}
+
+	if len(tasks) > pageSize {
+		tasks = tasks[:pageSize]
+		return tasks, tasks[len(tasks)-1].GetTaskKey().Next()
```
Comment (Shaddoll) [inline service/history/queuev2/queue_mem.go]: Can we just use the key of the dropped task?

File: service/history/queuev2/queue_mem.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+		copy(remaining, q.tasks[pos:])
+		q.tasks = remaining
+	}
+}
+
+// RTrimBySize truncates the queue to at most maxSize elements by dropping
+// the newest tasks from the right. Returns (newUpperBound, trimmed) where
+// newUpperBound is lastTask.GetTaskKey().Next() and trimmed indicates whether
+// any tasks were actually removed. Returns (MinimumHistoryTaskKey, false) if
+// the queue is empty after the operation, or (MinimumHistoryTaskKey, hadTasks)
```
Comment (Shaddoll) [inline service/history/queuev2/queue_mem.go]: before

File: service/history/queuev2/queue_mem.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+// if maxSize <= 0.
+func (q *inMemQueueImpl) RTrimBySize(maxSize int) (persistence.HistoryTaskKey, bool) {
```
Comment (Shaddoll) [inline service/history/queuev2/queue_mem.go:163]: The comment is confusing.

File: service/history/queuev2/queue_mem.go (hunk 2 above)
Comment (arzonus) [reply service/history/queuev2/queue_mem.go]: fixed, thanks!

File: service/history/queuev2/queue_mem.go (hunk 3 above)
Comment (arzonus) [reply service/history/queuev2/queue_mem.go:163]: fixed, thanks!

File: service/history/queuev2/queue_mem.go (hunk 1 above)
Comment (arzonus) [reply service/history/queuev2/queue_mem.go]: Changed to the dropped task's key, thanks!

**Review by Shaddoll — APPROVED** [review]
